### ISSUE TITLE
Return the actual query in one of the export files

### ISSUE THREAD

**Issue by ImkeF** [DaxStudio/DaxStudio#183]
It would be great if the actual query that has been executed would be returned in one of the output files that are generated when one saves the .dax-file.

In the special case where you select some rows in the script and execute just the selected rows, still the whole script is returned in the .dax-file.

I thought that this would be returned in the .dax.allQueries file (in the "Query"-record, but that always returns "EVALUATE ROW(\"BLANK\",0)". Not sure what to make out of this - maybe its actually a bug and should be returned here already?

However: In the "Query History"-tab in column "Query" stands what I'm looking for: The exact code that has been evaluated for the respective query. To see this in one of the exports would be very useful. Especially for presentations where one could then go through snippets of prepared code without having to manually (out-)comment them.

Thanks :)

**Comment by ImkeF** [comment]
Saw that this was an error on my side: Had all 3 tabs on, but just the server timings and query plan was actually recorded. Which produced the single "EVALUATE ROW("BLANK",0)" in the .allQueries-file.

So currently, the actual query executed is not returned anywhere when I'm on ServerTimings/QueryPlan-mode. Would be great to have that at least in the export files for later analysis.

**Comment by dgosbell** [comment]
> I thought that this would be returned in the .dax.allQueries file (in the "Query"-record, but that always returns "EVALUATE ROW("BLANK",0)". Not sure what to make out of this - maybe its actually a bug and should be returned here already?

The all queries trace should do this already. In addition to the "EVALUATE ROW("BLANK",0)" you should see the actual query you ran (sometimes it takes half a second or so for the trace event for your actual query to arrive). 

![image](https://user-images.githubusercontent.com/8241830/58755546-505df900-8529-11e9-97c9-d58c1339c3d7.png)

The "blank" query is an artifact of the "clear cache" command. The xmla ClearCache command does not just clear the data caches, it actually also clears the session and global scope too which is normally only established once per session. So DAX Studio runs this "blank" query which does not hit any data to force the session to be re-established. On big models with lots of measures and things like row level security this initializing of the session scope can be significant and you don't want these things included in the per query timings as they normally only happen once as the start of a session.

I could probably add a marker comment at the start of this blank query so that I could filter these out from the All Queries tab.

> Had all 3 tabs on, but just the server timings and query plan was actually recorded. 

This is a design limitation at the moment - you can't have all 3 tabs actively recording. We currently only run a single trace and just modify the events/filters for that trace. The server timings & query plan can run together as they require a trace filtered for the current session id. Where as the All Queries trace requires no session id filter so it is incompatible with the other 2 trace options.

We originally designed this around the scenario of running All Queries to capture a set of queries. Then switching that off and using server timings/query plans to tune individual slow queries.

I'm currently in the final weeks of helping organize our local SQL Saturday, so there is not much chance of getting time to look into working on a proper fix for this issue in that time, but in the mean time there are a couple of short term work arounds you could try.

## Short Term
* if you run *just* the All Queries trace you should get a list of all queries and their timings (including the "blank row" session warming queries)
* the query history is stored as a folder of json files under ```%appdata%\DaxStudio\QueryHistory``` you could use PowerQuery to bring these into an Excel table and if you have Server Timings running this should capture the Server Duration plus the FE and SE times. (and you could manually filter out the "blank row" queries and just filter for a specific date/time). The only downside is that we only keep a rolling history of the last 200 queries _(I have toyed with the idea of storing these in a SQLLite database or something like that so we could store more history and search it faster, but at the moment we've just kept it simple and lite)_

Once our SQL Saturday is done there are a few possible options I can think of that might be worth working on to help with this use case:

## Long Term
* I think it makes sense for All Queries to just start a dedicated trace of it's own so that it can run in parallel with the other two, but the query trace engine is one of the more complicated parts of the app. I think it also makes sense to filter out the "blank row" queries from the clear cache.
* It might be worth adding an option to allow for exporting the Query History to a single file
* I'm wondering if it might be useful to add a feature that could be turned on to "stamp" query stats comments underneath a query direct in the .dax file after you run it. I sometimes do this myself manually, I'm not sure if other people would find this useful or not.
eg.
```
//# DAX Studio Run 1 Duration: 321ms  Rows: 543
//# DAX Studio Run 2 Duration: 34ms  Rows: 543
```

**Comment by ImkeF** [comment]
Thanks so much Darren, very much appreciated!!
Yes, getting FE and SE-timings would exactly be what I need in addition to the current AllQueries-results. Unfortunately as it looks, these figures are not recorded while in AllQueries-mode (return -1 instead).

The workflow I had in mind was to analyze the results in Power BI and from there I would trigger an export via Python to csv for those sessions that I find worth saving. SE and FE figures would have been nice to have there - maybe they could even be integrated in the AllQueries table.

An alternative could be to trace the server timings in combination with Query Plan (only then the SE and FE timings are stored on my machine) and combine them with the appsource-data (as they contain the exact query that I'm after). But here the problem is that I don't get a field to match both tables: StartTime and Date created on appsource-data capture the times of the queries, while server timings capture the time when the file was stored. Working with index here is a slightly fragile option, as one has just to forget to save one run and the the history is messed up.

Just to say: This is absolutely not urgent for me, but it would be great to have it somewhere.
Thanks, Imke

BTW: Very much looking forward V 2.9. with integrated PBI Perf Data ;)
Dax Studio is an extremely helpful tool!

**Comment by marcosqlbi** [comment]
Recording FE and SE timings in AllQueries window is definitely too much expensive - believe me, you don't want to do that. Instead of capturing one event per query, you might captures tens to hundreds of events per query. In general, the right thing to do is: collect the queries, analyze execution time, when you spot something interesting dig deeper. A sistematic gathering of SE/FE is definitely too expensive. Microsoft should add some summarized metric in the Query End event, so we could do that without affecting server/engine performance.

**Comment by ImkeF** [comment]
Thanks Marco,
that aspect hasn't been on my radar.

Actually, I just found a way to match the query code from the appsource-data to the detailed timings-recordings: https://github.com/ImkeF/DAX-Tools/blob/master/DaxStudioAnalyzer_Timings.pbit 

That's actually all that I need at the moment. So thanks again Darren, for the hint!

**Comment by dgosbell** [comment]
> Yes, getting FE and SE-timings would exactly be what I need in addition to the current AllQueries-results. Unfortunately as it looks, these figures are not recorded while in AllQueries-mode (return -1 instead).

So if you have the Server Timings switched on, the information from that tab is passed to the Query History tab. _(the -1 show up as "n/a")_ 

![image](https://user-images.githubusercontent.com/8241830/58766985-ac2b8f00-85c8-11e9-8504-270094f61f5b.png)

Like Marco said, capturing Server Timings continuously or for more than just a single window could get expensive. But if you are running them in your session where you are testing your demo queries those stats will be captured.

**Comment by dgosbell** [comment]
@ImkeF If you are already using python in your template you could also use os.environ to automatically find the QueryHistory folder for the current user. The following M query should return the current user's Query History folder.

```
let
    #"Empty Table" = #table({"FolderPath"}, {{"a"}}),
    #"Run Python script" = Python.Execute("# 'dataset' holds the input data for this script#(lf)import os#(lf)fullpath = os.environ[""APPDATA""] + ""\DaxStudio\QueryHistory""#(lf)dataset[""FolderPath""].iloc[0] = fullpath",[dataset=#"Empty Table"]),
    dataset = #"Run Python script"{[Name="dataset"]}[Value],
    FolderPath = dataset{0}[FolderPath]
in
    FolderPath
```

**Comment by ImkeF** [comment]
Thanks @dgosbell , that's neat!
